### DISCUSSION TITLE
Viewing Views – Refresh / Reload

### DISCUSSION THREAD

**Discussion by ppii775** [mathesar-foundation/mathesar#911] [Maintainer discussions]
_This is the first of a series of new discussions for #466 **Support for viewing a database view**. This discussion is related to the now-closed discussions #624 and #653. Feel free to let me know if I missed anything or there's anything incorrect, since I'm new to the group and was not part of the last discussions. The Figma prototype that I'm using might contain outdated styles and components beyond the scope of this discussion._

## The Problem
> 1. The user should be able to see view data in tabular form, with columns and rows.
>         i. Error scenario: View data fails to load from the backend for some reason.

While working on the error states of loading a view, we started to think about the possible situations and how to let users retry. Besides unknown backend errors, error scenarios could be related to access privilege and collaboration, and GSuite has some good references. Please refer to **Example Error Scenarios** for details. If the loading errors were due to a simultaneous deletion or change of a view by another user, users might be provided an option to refresh the view panel. Please refer to **Questions Regarding Refresh and Save**. Some of these error scenarios and messages might apply to tables or Mathesar in general.

## Example Error Scenarios
**1. A view cannot be loaded due to access.**
<img width="1717" alt="Screen Shot 2021-12-20 at 9 51 18 PM" src="https://user-images.githubusercontent.com/6503097/146884283-4c6be798-e7ef-4e3c-98ce-4ecf95da9b51.png">

**2. A view cannot be loaded due to changes.**
When many people work on Google Slides and make changes simultaneously, a reload might be required. Not sure if this is something relevant to Mathesar.
<img width="1717" alt="Screen Shot 2021-12-20 at 10 12 43 PM" src="https://user-images.githubusercontent.com/6503097/146884780-cdc20f6a-083b-47e8-b30e-a1e9c68bedfe.png">

**3. A view cannot be loaded due to collaboration.**
<img width="1717" alt="Screen Shot 2021-12-20 at 10 13 50 PM" src="https://user-images.githubusercontent.com/6503097/146884829-541283a1-3971-4c70-bd1a-4ab58e49481f.png">

## Questions Regarding Refresh and Save
A view might not be able to be loaded if another user just deleted or modified the view. It is unclear how Mathesar handles changes in the background currently. Would it check for updates in a predetermined interval? Or would it provide a refresh/reload button for certain panels or the entire interface in general? 

**Possible solution 1:** A refresh status icon. This status icon would only show up and spin when updates are actively happening in the background. Alternatively, we could have a global refresh indicator. However, it is unclear where to put it on the current interface layout. Suggestions welcomed.
![image](https://user-images.githubusercontent.com/6503097/146885410-90f2093c-f970-43cd-b14c-10e7e4efde1e.png)

**Possible solution 2:** A refresh function in a view panel menu. It is unclear how Mathesar allows users to browse hundreds of views, besides doing a search. Is it possible to incorporate a sort or filter function for the view list as well?
![image](https://user-images.githubusercontent.com/6503097/146885622-aa252680-3b93-4e50-8993-f38ba73aaf30.png)

**Question regarding auto-save:** The refresh/reload function might work well with an auto-save function, unless we imagine users see changes in real time like Google Docs. Speaking of auto-save, it is unclear to me whether we have auto-save for the views, or maybe the status message "changes have been saved" would show up only after users click "save". It seems the tables in the prototype that I'm using have an auto-save function, but maybe that's just outdated. Nonetheless, is a combination of "auto-save" and "save as" for views a good option?

**Auto-save for tables?**
![image](https://user-images.githubusercontent.com/6503097/146886509-82f3ffe8-7d21-4fd3-b5bd-9b80afbcdae8.png)

**Manual save for views?**
![image](https://user-images.githubusercontent.com/6503097/146886133-37d1a602-2319-44f6-af60-ad146bba4cd1.png)

**Comment by kgodey** [comment]
We currently do not have any auto-save or background processing capability (and we don't plan to for the first version of Mathesar). The user would have to manually refresh the view. I think providing a way to manually refresh is a good idea.

> i. Error scenario: View data fails to load from the backend for some reason.

I should have been clearer when writing this; in this scenario, we do not know why the View data has failed to load. The backend may be down or produce an unknown error. The user might also have bad network connectivity.

I do think we might have future scenarios where the user has lost access or data has changed in the background, but we don't need to plan for those yet.

**Reply by ghislaineguerin** [reply]
@kgodey How would the user know when to refresh? My concern is that including a manual refresh button without a prompt could lead to the user becoming anxious about losing their progress or control over their work. Is there any way we can auto-refresh in the background? If not, could we associate the refresh to the save action? So that we confirm save and let users know we've also loaded all the latest changes. 
@ppii775 If we don't have auto-save, I'm thinking we'd need some way of letting users know they have unsaved changes or prevent refreshing if those changes aren't saved.

**Reply by kgodey** [reply]
> Is there any way we can auto-refresh in the background? If not, could we associate the refresh to the save action? So that we confirm save and let users know we've also loaded all the latest changes.

I think we should avoid auto-refresh for the initial release, that should be part of the [Real Time Collaboration](https://github.com/centerofci/mathesar/milestone/18) feature. We do always refresh when the user saves actions.

I'm not attached to the idea of having a refresh button.

**Reply by pavish** [reply]
@kgodey I think auto-save here refers to changes we do on the table like adding/deleting/updating rows. I may be mistaken, @ppii775 Is this what you're referring to?

If this is what you're referring to, we do infact auto-save without the user having to manually click a save button. This applies for tables currently and might apply for views. For views, adding a new row would probably involve a save button with the form, but there will not be a common save option on the top which saves a batch of changes. 

There will be no 'manual' save button which saves everything. Each individual change per user action gets saved whenever the user performs it.

**Reply by ppii775** [reply]
@pavish Yes I was referring to all user actions, for both tables and views. In the last section of the discussion I was trying to figure out what is the current auto-save capability that we are envisioning, since we seem to have auto-save for tables and manual save for views in the prototype that I'm using (though it could be outdated). If we in fact have this feature for tables, would it make sense to include for views for consistency? 

On the other hand, if we don't really need collaboration features for now as @kgodey mentioned, I think a reminder to save and a status symbol indicating unsaved changes should be sufficient.

In general, I would advocate for either making tables and views as similar as possible functionality-wise and look-wise, or making them more visually distinct (with the same design language) to emphasize the differences. As a non-technical user myself I can imagine the confusion being a potential issue.

**Reply by ppii775** [reply]
@pavish Thanks for sending over the staging server link! I actually have an additional question after checking out the latest version.

How are we imagining the auto-save and refresh button working together for tables right now (I assume they weren't built for collaboration)? Is refresh a way to get the latest data from DB outside of Mathesar? Will auto-save push updated data back to DB?

**Reply by pavish** [reply]
@ppii775 

> If we in fact have this feature for tables, would it make sense to include for views for consistency?

Yes, we would be having it for views as well. However, data addition from views work differently from tables. In tables, we'll be able to edit rows like a spreadsheet, where as editing in views would be more like a form submission, so we would have a save/add button there, but it would be for each single add/update action. We would not require a common save button. I guess we could call it semi-auto-save if that makes sense.

> How are we imagining the auto-save and refresh button working together for tables right now (I assume they weren't built for collaboration)? Is refresh a way to get the latest data from DB outside of Mathesar? Will auto-save push updated data back to DB?

Auto-save will update the DB directly. Refresh is there to explicitly fetch the latest content from the DB. As a note, every pagination page click or filter or any user action that fetches data would fetch latest data from DB at that point. Every other non-fetch user action (like adding, editing etc.,) would only perform that particular event and will not fetch latest data from DB.

Collaboration was not taken into account when we implemented the refresh button. Once we have collaboration or sync, we will have no need for explicit refresh since collaboration will involve the UI reflecting immediate changes from the DB.

**Comment by pavish** [comment]
Regarding the error messages, 

* A view cannot be loaded due to access

    We would only be displaying views that the user has access to, so this error should be rare. If access levels change while the user is using the product, I think a better place to show the error would be within the tab content instead of a modal. The user might be viewing a different table/view while the error occurs in one of the tabs.

* A view cannot be loaded due to changes. A view cannot be loaded due to collaboration.

    @kgodey has covered the collaboration aspect. These two errors are not something we need to worry about for the alpha. Currently, the most recent change made by any user is the change that gets updated on the db.

* We need an error message when connection drops with the backend.

    @kgodey mentioned it on her comment. This is a special type of error which is not specific to the view but the entire frontend. For this error, we can block all operations on the UI and show a modal with a error message. The modal will retry connecting to the backend in fixed intervals.

Regarding Refresh and Save,

* I think having a refresh option for the left-pane makes sense. We do infact already have a refresh option within each table.

**Reply by ppii775** [reply]
Thanks for your comments and clarifications!
> We would only be displaying views that the user has access to, so this error should be rare.

Is it possible that a user would be changing access privilege in real-time, causing access issues for a different user? I was also thinking about a collaboration scenario for this error, but if real-time collaboration is not a concern right now, I agree that it would be rare.

**Reply by kgodey** [reply]
> Is it possible that a user would be changing access privilege in real-time, causing access issues for a different user? I was also thinking about a collaboration scenario for this error, but if real-time collaboration is not a concern right now, I agree that it would be rare.

It is possible, yes.
